[Tooling] Add Buildkite annotation on test failures

AliSoftware opened the pull request:
This PR uses [Buildkite annotations](https://buildkite.com/docs/agent/v3/cli-annotate) to highlight the Unit and UI Tests that might have failed during a CI build.

![Example](https://user-images.githubusercontent.com/216089/176279184-19ce5eb4-9af5-438f-ab65-026d638008c5.png)

## To Test

 - If you see the build either fail on UI or Unit Tests because of flaky tests and those test failures being reported as annotations, that's great
 - If not and all test pass as they should (even the flaky ones), it's also a great sign that successful tests don't break (or try to annotate with an empty text, for example)
 - You can then use https://github.com/wordpress-mobile/WordPress-iOS/pull/18955 to test this PR by introducing test failure. In its current state, it not only introduces intentional fake failures in the tests, but also tweaks the pipeline to make it stub the junit file (useful to iterate faster when debugging this stuff!), but feel free to play with it to test more thoroughly.

## Alternatives Considered

This PR uses the route of using a simple XSLT to extract the failure cases from the JUnit report in a concise markdown text, which keeps the pipeline simpler.

I also tried implementing the same thing but using Buildkite's `junit-annotate` plugin in https://github.com/wordpress-mobile/WordPress-iOS/pull/18953, to compare the two approaches. After experimenting with that alternate approach, I realized that it created too much noise not only in the `pipeline.yml`, but also in the Build page (where the `wait` and `plugins: junit-annotate` additional steps are a bit "too" visible in the list of steps), and lastly required waiting for a new agent and spawning a docker container all that just for creating the annotation.

This is why in the end I found that the solution in this PR #18952 was not only simpler but also better and less resource-consuming.

## (ex-)Related PR

While working on this and discovering Buildkite's ability to group steps, I also played with it a bit to restructure the order of the steps in the pipeline. Those pipeline changes were initially built on top of this PR (hence outdated comments below), but I've since extracted them [in a separate PR #18954](https://github.com/wordpress-mobile/WordPress-iOS/pull/18954) (based on `trunk`) to isolate each unrelated changes in their own PR.

AliSoftware committed:
Add Buildkite annotation on test failures

AliSoftware committed:
🚧 Force dummy failures to test annotations

AliSoftware committed:
Improve the annotation formatting

AliSoftware committed:
Differentiate annotations per Buildkite step label

AliSoftware committed:
Switch XSLT to text output

To avoid inserted strings (like the test error messages) containing characters like < or > to be escaped in the output

AliSoftware commented:
Woohoo, look at [that](https://buildkite.com/automattic/wordpress-ios/builds/8617#0181abfc-fb57-49dd-b88e-e0535df5cb35):

<img width="1160" alt="image" src="https://user-images.githubusercontent.com/216089/176279184-19ce5eb4-9af5-438f-ab65-026d638008c5.png">

Managed to make it look [as neat as the plugin does](https://buildkite.com/automattic/wordpress-ios/builds/8605#0181ab76-fa58-402d-82a2-f359c522a24e), but with a simpler and less expensive solution (no separate and dependent step needed in pipeline, no separate agent spawned…) than [the one with the plugin](https://github.com/wordpress-mobile/WordPress-iOS/pull/18953). So I think this is the better solution 🙃 

I'm now gonna remove my debugging commits. See you in 1h once all the real build+test steps finally finish 😅

AliSoftware commented on .buildkite/pipeline.yml:
```diff
@@ -64,6 +64,33 @@ steps:
+  #################
+  # UI Tests
+  #################
+  - group: "🔬 UI Tests"
```
Took the occasion of this PR cleaning up how the Build page looks, by also:

 - Moving those two "UI Test" steps _before_ the "Lint Translations" step, so that they are next to the "Build for Testing" and that all the related "Run {Unit|UI} Tests" steps stay together
 - Switching to [use a Group step](https://buildkite.com/docs/pipelines/group-step) to group the two related UI Tests together, making the list of steps shown in the Build header/summary look neater. I thought about putting the Unit Test in the same logical group, but figured it would probably be too much grouping, while grouping the 2 steps calling the exact same `.sh` script but for iPhone vs iPad made more sense.

<img width="1157" alt="image" src="https://user-images.githubusercontent.com/216089/176284306-cf4b70b0-a25b-4865-8700-fcb519f59949.png">

AliSoftware replied:
Just realized there's an opportunity to similarly group the 2 "Installable Build" steps together. Will push a commit doing that as well later.

AliSoftware replied:
Ended up deciding to separate that change (pipeline reorg) in a dedicated PR, to avoid polluting this one and to keep the two different changes separate.

This part is now implemented in https://github.com/wordpress-mobile/WordPress-iOS/pull/18954, which is now a fully independent PR (based on `trunk`, and not stacked on top of this one anymore)

AliSoftware committed:
XSLT codestyle cleanup

AliSoftware commented:
Mmmh ok strangely it seems when we run the real steps the command exits with code 1…

But [when I strip the `pipeline.yml` to skip the real build, and test and replace them with dummy steps which just write the `build/results/report.junit` file](https://github.com/wordpress-mobile/WordPress-iOS/pull/18955) — which should produce the same thing because the `report.junit` file I place here in those dummy steps is one I downloaded from the artefacts of an actual previous failed build… — it worked correctly 🤔 

---

[EDIT] Ok, I think I know why: the difference being in the real case, we do a command to `cd build/results/ && zip …` to zip the `xcresult` bundle… but don't revert the PWD back after that. So that means we were probably in the wrong working directory when we tried to run `xsltproc`. What still feels strange is that my `if [ -f "$JUNIT_REPORT" ]` check should have guarded against the command to even run, as if we were indeed in the wrong directory I'd expect that the relative path would not be found and `-f` to be `false`… so still unsure and thus only half-confident my test is correct here then…

Let's see if 25e8b00 fixes it anyway…

AliSoftware committed:
Revert to previous directory after `zip`

Otherwise it breaks the next step because we're not in the $PWD we expect anymore when trying to push the Buildkit annotation

mokagio commented on .buildkite/commands/run-ui-tests.sh:
```diff
@@ -47,5 +47,10 @@ if [[ $TESTS_EXIT_STATUS -eq 0 ]]; then
+  JUNIT_REPORT="build/results/report.junit"
+  if [ -f "$JUNIT_REPORT" ]; then
+    xsltproc --stringparam step_title "$BUILDKITE_LABEL" ".buildkite/commands/junit-failures-to-buildkite-annotation.xslt" "$JUNIT_REPORT" | buildkite-agent annotate --style error --context "ui-tests-$TEST_NAME-$DEVICE"
+  fi
```
@AliSoftware, you [commented](https://github.com/wordpress-mobile/WordPress-iOS/pull/18952/#issuecomment-1169332988):

> the difference being in the real case, we do a command to cd build/results/ && zip … to zip the xcresult bundle… but don't revert the PWD back after that.

I can see the `&& cd -` in the unit tests command script via [`25e8b00`](https://github.com/wordpress-mobile/WordPress-iOS/pull/18952/commits/25e8b0063a0057bf3a1c596d0947444fcab3e4b3), but I can't find one here, in the UI tests command ([see line 42](https://github.com/wordpress-mobile/WordPress-iOS/pull/18952/files#diff-bd954ba2ef49cba9cd9950b5197be8e22f9c95f5482f03d0186eaf9404cda0e8R42)).

AliSoftware replied:
Dammit, good point, forgot to add it here as well, nice catch!

@mokagio Can I let you add it, then tweak https://github.com/wordpress-mobile/WordPress-iOS/pull/18955 in your day to trigger a build that would test the case of when there's an actual UI Test failure (basically, probably only keep 63e73e8037b91f883657547743bd707ea1ea1f05 in that test PR, but `git revert` bc15c6c52ab8ffa8f9f9a0c3f081efa0448601ad, to still run the full pipeline this time, but with intentional test failures)? 🥺 🙏 

That should allow us to validate that this `&& cd -` was really what was missing and that it solves the previous `exit 1` error when test failure occurs, in both Unit and UI tests

mokagio replied:
Done, the build is running here: https://buildkite.com/automattic/wordpress-ios/builds/8636#0181ad9f-9761-408c-823e-72812b8951ed

Leaving this WIP comment in case I don't get a chance to see the build finish and wrap this up.

AliSoftware replied:
Thanks, seems it worked! 🎉 

<img width="1162" alt="image" src="https://user-images.githubusercontent.com/216089/176398447-e120fe31-953d-4799-9c92-c106fce87bf0.png">

Still unsure why, when we were not in the right PWD before with the missing `cd -`, the `if [ -f "$JUNIT_REPORT" ]` didn't fail and prevent the command from crashing…

mokagio reviewed (commented):
Great initiative, thank you very much!

I added it to the iOS Core project because it's something that fits in our vision to remove friction in the dev cycle :tada: 🙇‍♂️

mokagio committed:
Revert to previous directory after `zip`, for UI tests as well

Same as the previous commit, 25e8b0063a0057bf3a1c596d0947444fcab3e4b3,
but for the Buildkite command that runs the UI test.

> Otherwise it breaks the next step because we're not in the `$PWD` we
> expect anymore when trying to push the Buildkite annotation

AliSoftware committed:
Use new annotate_test_failures from bash-cache

AliSoftware commented:
@mokagio Since the fix seems to have worked as expected, this should now be ready to be approved and merged.

That being said, in order to make this usable by other repos, I've started to create [bash-cache-buildkite-plugin#27](https://github.com/Automattic/bash-cache-buildkite-plugin/pull/27) + [#18961](https://github.com/wordpress-mobile/WordPress-iOS/pull/18961) to migrate this into the `bash-cache` plugin.

Not sure if you prefer to:
1. Merge this PR as is, then reviewing the 2 new PRs (bash-cache and WPiOS) separately as an improvement over this one
2. Or wait for those 2 new PRs to be approved and merged — which will require us to do a new release of the plugin — and merge the final result all at once

I've built [#18961](https://github.com/wordpress-mobile/WordPress-iOS/pull/18961) on top of this PR's branch, so we can go either route: either (1) merge this first, then merge the 2 PRs later after, or (2) merge the 2 new PRs first into this one and merge this PR to land everything at once into `trunk`. Up to you, both are fine with me either way!

AliSoftware committed:
Revert "🚧 Force dummy failures to test annotations"

This reverts commit da8271e7b5dacb9399b737f1de156fed8a7d518c.

AliSoftware committed:
`annotate_test_failures` now checks the file exists

AliSoftware committed:
Call `annotate_test_failures` even on success

So that it can remove any potential previous test failure annotation in case of flaky tests that failed on first run but ends up passing after a retry.

AliSoftware committed:
Update bash-cache plugin version

With latest release including the new command we use here

AliSoftware committed:
Merge pull request #18961 from wordpress-mobile/tooling/annotate-test-failures-toolkit

[Tooling] Add Buildkite annotation on test failures — DRY using bash-cache plugin

AliSoftware commented:
Admin-merging this because https://github.com/wordpress-mobile/WordPress-iOS/pull/18961, which was a replacement for this, was since approved, and this current PR is basically landing all this already-approved diff to `trunk` by now 🙃
